tools/runbook-sync-downstream/runbook.go

```diff
@@ -0,0 +1,192 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/grafana/regexp"
+	"k8s.io/klog/v2"
+)
+
+var (
+	runbookRegex = regexp.MustCompile(`.*\.md`)
+
+	editedByRegex           = regexp.MustCompile(`<!-- Edited by .*-->`)
+	downstreamCommentsRegex = regexp.MustCompile(`<!--DS: (.*)-->`)
+	multipleNewLinesRegex   = regexp.MustCompile(`\n\n+`)
+	asciiDocLinksRegex      = regexp.MustCompile(`link:(https://[^[\]]+)\[([^[\]]+)\]`)
+)
+
+type runbook struct {
+	name string
+
+	lastLocalUpdate     time.Time
+	upstreamLastUpdated time.Time
+}
+
+func listRunbooksThatNeedUpdate(downstreamRepo *git.Repository, upstreamRepo *git.Repository) ([]runbook, []runbook) {
+	localRunbooks, err := findRunbooksLastUpdateDates(downstreamRepo, downstreamRunbooksDir)
+	if err != nil {
+		klog.Fatal(fmt.Errorf("failed to find runbooks last update dates: %w", err))
+	}
+
+	upstreamRunbooks, err := findRunbooksLastUpdateDates(upstreamRepo, upstreamRunbooksDir)
+	if err != nil {
+		klog.Fatal(fmt.Errorf("failed to find runbooks last update dates: %w", err))
+	}
+
+	return checkWhichRunbooksNeedUpdate(localRunbooks, upstreamRunbooks)
+}
+
+func findRunbooksLastUpdateDates(repo *git.Repository, dir string) (map[string]time.Time, error) {
+	runbooksTree, err := getDirCurrentTree(repo, dir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current runbooks tree: %w", err)
+	}
+
+	runbooks := make(map[string]time.Time)
+	for _, entry := range runbooksTree.Entries {
+		if !runbookRegex.MatchString(entry.Name) {
+			continue
+		}
+
+		lastCommitDate, err := findLastCommitDate(repo, dir, entry.Name, nil)
+		if err != nil {
+			return nil, fmt.Errorf("failed to find last commit date: %w", err)
+		}
+		runbooks[entry.Name] = lastCommitDate
+	}
+
+	return runbooks, nil
+}
+
+func checkWhichRunbooksNeedUpdate(localRunbooks, upstreamRunbooks map[string]time.Time) ([]runbook, []runbook) {
```

Comment:
Please consider splitting this function to two functions, one for listing runbooks to update, and another for listing runbooks to delete. IMO it would be a bit more readable.

Reply:
updated